lib/index.js

```diff
@@ -1,55 +1,82 @@
-        gzip: {
-            level: 6
-        },
-        overwrite: false
-    };
+const zlib = require('zlib'),
+      bytes = require('bytes');
+
+function compress(file, path, gzipOpts) {
+    return new Promise((resolve, reject) => {
+        const gzip = zlib.createGzip(gzipOpts);
+        const compressedChunks = [];
 
-    if (typeof options === 'object') {
-      options = Object.assign(defaultOptions, options);
-    } else {
-        options = defaultOptions;
-    }
+        gzip.on('data', function storeCompressedChunk(chunk) {
+            compressedChunks.push(chunk);
+        });
 
-    return function metalsmithGzip(files, metalsmith, done) {
+        gzip.on('error', reject);
 
-        const filesTbCompressed = mm(Object.keys(files), options.src);
+        gzip.on('end', function concatenateCompressedChunks() {
+            resolve([path, Buffer.concat(compressedChunks)]);
+        });
+
+        gzip.write(file.contents);
+        gzip.end();
+    });
+}
 
-        each(filesTbCompressed, function compress(file, asyncDone) {
-            const gzip = zlib.createGzip(options.gzip);
-            const data = files[file];
-            const compressedChunks = [];
+/**
+ * @typedef {Object} Options
+ * @property {string} [src] Glob pattern of files to match for compression
+ * @property {import('zlib').ZlibOptions} [gzip] Options to pass to `zlib.createGzip`
+ * @property {boolean} [overwrite=false] Whether to overwrite original files (true) or create a new file with .gz suffix.
+ */
 
-            gzip.on('data', function storeCompressedChunk(chunk) {
-                compressedChunks.push(chunk);
-            });
+/** @type {Options} */
+const defaultOptions = {
+    src: "**/*.+(html|css|js|json|xml|svg|txt)",
+    gzip: {
+        level: 6
+    },
+    overwrite: false
+};
 
-            gzip.on('error', function signalError(error) {
-                asyncDone(error);
-            });
+/**
+ * A Metalsmith plugin to compress build files with gzip.
+ * @param {Options} options 
+ * @returns {import('metalsmith').Plugin}
+ */
+function gzip(options) {
+    options = { ...defaultOptions, ...(options || {}) }
 
-            gzip.on('end', function concatenateCompressedChunks() {
-                if (options.overwrite) {
-                    data.contents = Buffer.concat(compressedChunks);
-                } else {
-                    const compressedFile = file + '.gz';
-                    files[compressedFile] = Object.assign({}, data);
-                    files[compressedFile].contents = Buffer.concat(compressedChunks);
-                }
-                asyncDone();
-            });
+    return function gzip(files, metalsmith, done) {
+        const debug = metalsmith.debug('metalsmith-gzip')
+        debug('Running with options %o', options)
 
-            gzip.write(data.contents);
-            gzip.end();
+        const filesTbCompressed = metalsmith.match(options.src, Object.keys(files));
```

Comment:
Pretty smart to have added `match()` to the Metalsmith API

Reply:
Haha yeah, was my first addition when I took over